### DISCUSSION TITLE
LuaFAR for Far2L

### DISCUSSION THREAD

**Discussion by shmuz** [elfmz/far2l#1197] [Show and tell]
Создал репозиторию для LuaFAR и Lua-плагинов (для Far2L): https://github.com/shmuz/LuaFAR-2L
На данный момент в ней есть исходники луафар и двух плагинов.
У меня это работает на Linux Mint 64 bit.
Если кто не знает, что всё это такое, можно посмотреть файлы документации. Для тех, кто знаком с Far3, наверное уже понятно.
Если кто-то хочет бинарники для Linux Mint 64 bit и совместимых, могу выложить.
В Линуксе я начинающий, со всеми вытекающими последствиями.

**Comment by unxed** [comment]
Я просто оставлю это здесь
https://forum.farmanager.com/viewtopic.php?p=169614#p169614

**Comment by unxed** [comment]
> Создал репозиторию для LuaFAR и Lua-плагинов (для Far2L): https://github.com/shmuz/LuaFAR-2L На данный момент в ней есть исходники луафар и двух плагинов. У меня это работает на Linux Mint 64 bit. Если кто не знает, что всё это такое, можно посмотреть файлы документации. Для тех, кто знаком с Far3, наверное уже понятно. Если кто-то хочет бинарники для Linux Mint 64 bit и совместимых, могу выложить. В Линуксе я начинающий, со всеми вытекающими последствиями.

Штука, которая начинающим не очевидна: в разных версиях дистрибутивов разные версии libc, так что фар из минт 20 не взлетит в минт 19. Так что когда речь о бинарниках, надо писать точную версию дистра, в котором собирали

PS круто что портировали lua, спасибо!

**Comment by shmuz** [comment]
С помощью гугла выполнил 2 команды, этого достаточно или нет?
```
$ cat /etc/os-release
NAME="Linux Mint"
VERSION="20.2 (Uma)"
---------------------------------------------------------------------------
$ ldd --version
ldd (Ubuntu GLIBC 2.31-0ubuntu9.2) 2.31
```

**Comment by unxed** [comment]
Достаточно! :)

**Comment by spvkgn** [comment]
> Если кто-то хочет бинарники для Linux Mint 64 bit и совместимых, могу выложить.

Было бы отлично, ибо как из собрать самому непонятно

**Comment by shmuz** [comment]
@spvkgn, здесь: https://github.com/shmuz/LuaFAR-2L/releases
Не забудьте (если не было раньше сделано): ```sudo apt install lua5.1```
И если не работает или плохо работает, не стесняйтесь сообщать.

**Reply by unxed** [reply]
`sudo apt install liblua5.1-0` ещё надо

**Comment by unxed** [comment]
А можно этот lua-плагин портировать для far2l, пожалуйста?

https://forum.farmanager.com/viewtopic.php?p=113823#p113823

**Reply by shmuz** [reply]
С большой вероятностью можно, но пока не попробуешь, наверняка не узнаешь.
Но у меня приоритеты такие, по порядку:
1. портирование luafar - оно хотя и  почти закончено, но совершенно не обкатано
2. портирование *своих* плагинов и скриптов

Так что до портирования чьих-то скриптов ещё должно пройти время, да и надеюсь что будет кроме меня кому этим заняться.

**Reply by unxed** [reply]
Ага, спасибо! Это не очень подгорает. Хотя было бы приятно! Там 187 строк, из которых треть — комментарии, по идее, не очень сложный он

**Comment by unxed** [comment]
И ещё, можно, пожалуйста, подсказочку, как в Mint плагин собрать? Хочу сам попробовать, а то готовый бинарник не заводится чот

**Reply by shmuz** [reply]
А у вас Mint 20 64 бит?
Я перед выкладыванием релиза проверил, что всё заводится - распаковал релиз в специально созданную копию far2l и запустил.
Lua 5.1 у вас установлен? (```sudo apt install lua5.1```)
И расскажите, куда вы распаковываете архив релиза.
А инструкции по сборке я через какое-то время сделаю, там надо наводить порядок.

**Reply by unxed** [reply]
$ cat /etc/lsb-release
DISTRIB_ID=LinuxMint
DISTRIB_RELEASE=20.3
DISTRIB_CODENAME=una
DISTRIB_DESCRIPTION="Linux Mint 20.3 Una"
$ uname -m
x86_64

Lua ставил именно этой командой только что.
Распаковываю в /usr/lib/far2l, поверх папки Plugins.

Угу, инструкции по сборке это очень здорово было бы, спасибо! Я ни в коем случае не тороплю и понимаю, что это всё время занимает :) Просто поиграться интересно, а никак))

**Reply by shmuz** [reply]
После распаковки папки `_luafar_`, `_lf4ed_` и `_lfsearch_` должны лежать рядом с папками стандартных плагинов. У вас так?
Если так, и всё же в меню плагинов не появляются новые пункты, то буду думать.
А сообщения об ошибках от far2l при запуске есть?

**Reply by shmuz** [reply]
Пожалуйста сделайте ldd на всех бинарниках из установленного релиза и покажите мне вывод:
```
$ ldd ./luafar2l.so
$ ldd ./lf4ed-x64.far-plug-wide
$ ldd ./lfsearch-x64.far-plug-wide
```

**Reply by unxed** [reply]
$ ldd ./lf4ed-x64.far-plug-wide
        linux-vdso.so.1 (0x00007ffd94b0c000)
        /home/shmuel/far2l/_build/install/Plugins/_luafar_/src/Out64/luafar2l.so => not found
$ ldd ./lfsearch-x64.far-plug-wide
        linux-vdso.so.1 (0x00007fff93d70000)
        /home/shmuel/far2l/_build/install/Plugins/_luafar_/src/Out64/luafar2l.so => not found
$ ldd ./luafar2l.so
        linux-vdso.so.1 (0x00007fff4b7c1000)
        liblua5.1.so.0 => not found
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ff59abb3000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ff59ae17000)
$ sudo apt install lua5.1
[sudo] пароль для unxed:
Чтение списков пакетов… Готово
Построение дерева зависимостей
Чтение информации о состоянии… Готово
Уже установлен пакет lua5.1 самой новой версии (5.1.5-8.1build4).
Обновлено 0 пакетов, установлено 0 новых пакетов, для удаления отмечено 0 пакетов, и 0 пакетов не о
бновлено.

**Reply by unxed** [reply]
Кажется, вот чего надо было сделать
`sudo apt install liblua5.1-0`

**Reply by unxed** [reply]
Стал запускаться luafar, но не взлетает остальное:

```
$ ldd ./luafar2l.so
        linux-vdso.so.1 (0x00007fff0e56a000)
        liblua5.1.so.0 => /lib/x86_64-linux-gnu/liblua5.1.so.0 (0x00007ffbdeafd000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ffbde90b000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ffbde7bc000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007ffbde7b6000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ffbdeba0000)
$ ldd ./lfsearch-x64.far-plug-wide
        linux-vdso.so.1 (0x00007ffd585c5000)
        /home/shmuel/far2l/_build/install/Plugins/_luafar_/src/Out64/luafar2l.so => not found
$ ldd ./lf4ed-x64.far-plug-wide
        linux-vdso.so.1 (0x00007ffc25e96000)
        /home/shmuel/far2l/_build/install/Plugins/_luafar_/src/Out64/luafar2l.so => not found

```

Кажется, там жёстко вкомпилён путь /home/shmuel/far2l/_build/install/Plugins, а надо, чтоб был /usr/lib/far2l/Plugins

**Reply by shmuz** [reply]
@unxed, в течение ближайшего часа выложу новый релиз, там не будет зависимости от `/home/shmuel`

**Reply by shmuz** [reply]
Выложил. После распаковки надо зайти в папку `_luafar_` и запустить `install.sh`.
После чего перезапустить фар.

**Comment by elfmz** [comment]
Путь надо определять через то что приходит в PluginModuleOpen, более того, то что приходит в PluginModuleOpen - это путь к бинарнику плагина, то есть в /usr/lib/far2l/Plugins, а скрипты это не бинарники и потому то что приходит в PluginModuleOpen следует прогнать через функцию из utils - TranslateInstallPath_Lib2Share(pluginPath); - и получится путь типа /usr/share/far2l/Plugins - это все можно подсмотреть в плагине python.
А если совсем по хорошему то надо продумать еще дополнительный путь для местонахождения скриптов гденить в юзерском профиле, то есть в ~/.config/far2l/plugins/lua... - этот путь можно получить через InMyConfig("plugins/lua...")

ЗЫ Самое большое различие между программированием под винду и под юниксы - под винду вы пишете под конкретное окружение (ось, компилятор, либы, конфигурация) под юникс пишете сразу под зоопарк.

**Reply by shmuz** [reply]
@elfmz, мне потребуется время, чтобы понять эту информацию.
Пока что не до оптимизаций, забил путь установки luafar2l.so в `/usr/lib/x86_64-linux-gnu/`,  а впоследствии можно будет это изменить.

**Reply by unxed** [reply]
Круто — кажется, теперь запускается! А как конкретный скрипт выполнить? Где в интерфейсе lua искать?

**Reply by shmuz** [reply]
1. В панелях, в меню плагинов должен появиться пункт LuaFAR Search
2. В редакторе, в меню плагинов должны появиться 2 новых пункта
3. Если скрипт (например) test.lua находится в текущей папке, то можно запустить `lfe: -r test.lua`
Для большей информации почитайте `lf4ed_manual.chm`

**Reply by shmuz** [reply]
@unxed, если не появились новые пункты в меню плагинов, повторите снова `ldd` для всех `far-plug-wide`

**Reply by unxed** [reply]
> @unxed, если не появились новые пункты в меню плагинов, повторите снова `ldd` для всех `far-plug-wide`

Не появились :(

...far2l/lua/LuaFAR-2L/Plugins/_lfsearch_/plug$ lfe: -r ./_usermenu.lua
lfe:: команда не найдена

/home/unxed/far2l/lua/1234$ ldd ./luafar2l.so
        linux-vdso.so.1 (0x00007ffcfad55000)
        liblua5.1.so.0 => /lib/x86_64-linux-gnu/liblua5.1.so.0 (0x00007fdb37d4c000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fdb37b5a000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fdb37a0b000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fdb37a05000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fdb37def000)

/home/unxed/far2l/lua/1234$ ldd ./lfsearch-x64.far-plug-wide
        linux-vdso.so.1 (0x00007ffc92f54000)
        /usr/lib/x86_64-linux-gnu/luafar2l.so (0x00007f6190fa9000)
        liblua5.1.so.0 => /lib/x86_64-linux-gnu/liblua5.1.so.0 (0x00007f6190f55000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f6190d63000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f6190c14000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f6190c0e000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f619101a000)

/home/unxed/far2l/lua/1234$ ldd ./lf4ed-x64.far-plug-wide
        linux-vdso.so.1 (0x00007ffff6cc7000)
        /usr/lib/x86_64-linux-gnu/luafar2l.so (0x00007fcb2ab1c000)
        liblua5.1.so.0 => /lib/x86_64-linux-gnu/liblua5.1.so.0 (0x00007fcb2aac8000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fcb2a8d6000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fcb2a787000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fcb2a781000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fcb2ab85000)

**Reply by shmuz** [reply]
Ох, даже не знаю что сказать, трудно быть новичком.
У меня с вами разница в том, что у вас liblua5.1 установлена, начиная от /lib а у меня начиная от /usr/lib. Но не думаю, что это влияет.
- У вас папка `_lf4ed_` лежит рядом с папками стандартных плагинов (colorer и проч.) ?
- Фар перезапускали?
- Фар выдаёт ошибки при запуске?

**Reply by unxed** [reply]
```
/usr/lib/far2l/Plugins$ ls /usr/lib/far2l/Plugins/
align     calc     compare   editcase    filecase  inside   _lfsearch_  multiarc  SimpleIndent
autowrap  colorer  drawline  editorcomp  incsrch   _lf4ed_  _luafar_    NetRocks  tmppanel
```
Перезапускал. Не выдаёт.

**Reply by shmuz** [reply]
А если вы в Фаре или в терминале запускаете просто `lua`, что происходит?

**Reply by unxed** [reply]
> А если вы в Фаре или в терминале запускаете просто `lua`, что происходит?

/home/unxed$ lua
Lua 5.1.5  Copyright (C) 1994-2012 Lua.org, PUC-Rio
\>

**Reply by shmuz** [reply]
OK, спасибо. Это всё временные проблемы, они будут решены раньше или позже.
Возможно соберу для вас дебаг-версию плагина, она сделает лог вызовов от Фара и тогда всё увидим.
Но это не сию минуту. Если через несколько дней ничего не изменится, соберу. Пока буду двигаться дальше.

**Reply by unxed** [reply]
Может, я не там ищу? В окне, которое по F11, Lua нет. В окне, которое по Alt-F1 — тоже. И в редакторе по F11 нету.

**Comment by shmuz** [comment]
Вы там ищете - в окне по F11.
Не беспокойтесь, проблему эту обязательно решим, только несколько позже. Я в этом весьма заинтересован.

**Reply by unxed** [reply]
Запустил far2l в отладочном режиме (`FAR2L_STD=~/far2l.log far2l`)
Там такое:
```
LoadPlugin: cache hit for '/usr/lib/far2l/Plugins/_lfsearch_/plug/lfsearch-x64.far-plug-wide'
LoadPlugin: cache hit for '/usr/lib/far2l/Plugins/_lf4ed_/plug/lf4ed-x64.far-plug-wide'
```
В `/home/unxed/.config/far2l/plugins/state.ini` они оба тоже фигурируют. То есть вроде бы он их откртывает и запускает, но дальше почему-то ничего не происходит.

**Reply by shmuz** [reply]
@unxed, тут дело в так называемых permissions (в которых я пока не разбираюсь).
Когда я скопировал Фар в `usr/lib` (с помощью sudo), затем скопировал папки `_lf4ed_` и `_lfsearch_`, и затем запустил эту копию Фар, то во первых мои плагины не взлетели, а во вторых система даже не даёт мне зайти в вышеуказанные папки (а в папки стандартных плагинов - пожалуйста).

Если вы знаете как эта проблема решается - расскажите.
А иначе можете установить копию фара где-нибудь под `home/unxed` и думаю, что мои плагины появятся.
Только не копируйте их из usr/lib/far2l, иначе будет та же история.

**Reply by unxed** [reply]
Попробовал из /home/unxed, та же история :( Выставил на *.far-plug-wide и *.so такие же разрешения, что и для других плагинов стоят — не срабатывает всё ещё. Что-то мы упускаем

**Reply by shmuz** [reply]
Вот инструкции по сборке:
Заходите в исходники (не в инсталляцию)

1. в папке `_luafar_/src`:
`sudo make install`
2. в папке `_lf4ed_/plug/build`:
`make embed` - бинарник появится в папке `plug`
3. в папке `_lfsearch_/plug/build`:
`make -fembed.mak` - бинарник появится в папке `build`

С пунктом (1) больше ничего не надо делать (уже установлено), а бинарники из пунктов (2) и (3) скопируйте на место старых в инсталляцию.
P.S. Обязательно, чтобы сорсы соответствовали релизу, т.к. API сейчас каждый день меняется.

**Reply by unxed** [reply]
1) исходники far2l надо клонировать из git прямо в home, потому что в мейкфайлах жёстко пути
прописаны. Поэтому только так:
```
cd ~
git clone https://github.com/elfmz/far2l
```

`~` это в *nix синоним домашней папки, если что.

2) Исходники релиза Lua-FAR2l надо класть сюда:
`~/far2l/_build/install/Plugins/`

3) Перед сборкой нужно сделать `sudo apt install liblua5.1-dev`, иначе ld не находит библиотеки

4) `sudo make install` в `_luafar_/src` не отработало:
```
$ sudo make install
[sudo] пароль для unxed:
make: *** Нет правила для сборки цели «/root/far2l/far2l/Include/plugin.hpp»,
требуемой для «farflags.c».  Останов.
```

Но отработал обычный `make`.

NB! В `luafar2l.so` всё ещё встречается захардкоженный путь:
`/home/shmuel/far2l/_build/install/Plugins/_lfsearch_/mylog.txt`
Я там PR отправил, исправляющий это.

5) Дальше копируем `luafar2l.so` в `/usr/lib/x86_64-linux-gnu/`

6) Запускаем far2l из `~/far2l/_build/install`, и

Урааа, у меню F11 появляется LuaFAR for Editor!

**Reply by unxed** [reply]
Но там же, в меню F11, должно три новых пункта появиться? Или два? Я пока только один вижу. Однако reformat block в редакторе сделать удалось, то есть по крайней мере что-то уже работает, класс!

**Reply by shmuz** [reply]
@unxed, хорошие новости! Так что получается, бинарники от Mint 20.2 не совместимы с 20.3 ?
В меню в панелях должен появиться LF Search, в меню в редакторе - он же + LF for Editor
(сейчас я занят по работе, освобожусь ближе к вечеру).

**Reply by unxed** [reply]
> @unxed, хорошие новости! Так что получается, бинарники от Mint 20.2 не совместимы с 20.3 ? В меню в панелях должен появиться LF Search, в меню в редакторе - он же + LF for Editor (сейчас я занят по работе, освобожусь ближе к вечеру).

От 20.2 по идее должны быть совместимы с 20.3. Если бы были не совместимы — far2l ругался бы, что загрузить не может. Там что-то другое, не понял, что, пока. Ну, по крайней мере, хотя бы один взлетел, уже классно!

Остальное понемногу докрутим, не горит, конечно :)

**Reply by shmuz** [reply]
@elfmz 
> Путь надо определять через то что приходит в PluginModuleOpen

Пытался найти описание `PluginModuleOpen` в гугле, в хелпе far2l и в plugin.hpp - потерпел неудачу.
Вижу его в плагине python.
Может быть где-то есть какая-то документация по far2l, где это (и не только) описано, кроме как читать исходники far2l ?
Если есть, дайте ссылочку, пожалуйста.

**Reply by elfmz** [reply]
PluginModuleOpen - это просто функция которую экспортирует плагин и которую far2l вызывает при загрузке модуля плагина (если она есть). По сути аналог DllStartup, но сделанный в рамках far2l. Ему в аргументах приходит путь к загруженному бинарнику плагина, от которого уже можно кудато отталкиваться.

**Reply by shmuz** [reply]
Спасибо. Давайте добавим её прототип в plugin.hpp рядом с другими экспортируемыми функциями.
Если надо пул-реквест - сделаю.

**Reply by elfmz** [reply]
Ок позже добавлю, на заметку - plugin.hpp pluginold.hpp и вся папка Include - это legacy, не легаси это farplug-mb.h и farplug-wide.h в far2sdk. Причина такого внезапного равноправия - в far2l оба вида плагинов юникодные, просто mb - это UTF8, а wide - UTF32. В идеале хочется оба варианта сделать полностью одинаковыми, за исключением формата строк.

**Reply by elfmz** [reply]
Еще сразу можете глянуть как в питоне сделана асинхронная инициализация плагина, чтоб минимизировать влияние на время запуска фара. ThreadProc - это процедура потока, запускаемая StartThread()-ом сразу при загрузке и которую ожидают по мере необходимости

**Comment by shmuz** [comment]
@elfmz , это всё очень значительная для меня информация, и совершенно новая. Если её нигде не существует в письменном виде, буду собирать здесь понемногу. Спасибо.

**Reply by elfmz** [reply]
ну переименование инклудников - это на самом деле новая вещь, сделано было непосредственно перед бетой, хотя хотелось сделать это очень давно, но руки не доходили просто

**Comment by shmuz** [comment]
У меня в заголовке показывает "alpha" (far2l собран месяц назад). Уже стал бетой? (Хотя это всё условности).

**Reply by elfmz** [reply]
уже 10 дней как бета) условности, да

**Comment by unxed** [comment]
к слову об удобной сборке. я вот когда портировал простейший плагин, сразу собирал с помощью cmake, чтоб удобно было. вот, @shmuz  гляньте на его cmake-скрипт, может, получится по аналогии сделать для lua

https://github.com/unxed/far2l-EditWrap/blob/main/CMakeLists.txt

**Reply by shmuz** [reply]
@unxed , да, знаю, что cmake это мощная штука, но мне ещё не довелось с ней познакомиться.
Может быть, когда-нибудь. Для моих поделок покуда обычного make хватает.

**Comment by shmuz** [comment]
@unxed , я изменил структуру папок и мейкфайлы.
luafar2l.so больше не устанавливается в /usr/lib/x86_64-linux-gnu, но тем не менее должно работать.
Буду признателен, если проверите, когда будет время (естественно, все 3 бинарника надо будет пересобрать).

**Reply by unxed** [reply]
Всё собралось. В меню по F11 теперь есть два пункта — «LuaFAR for Editor» и «LuaFAR Search». В редакторе — тоже. А вот в Alt+F1 ничего не добавилось.

**Reply by unxed** [reply]
LuaFAR Search при запуске падает, правда

**Comment by shmuz** [comment]
В Alt-F1/F2 плагины ничего не добавляют (это просто делается, но там и без них всего много).
А насчёт падения, есть ли сообщение на экране, и если да, то что там?
Да, и желательно удалить `/usr/lib/x86_64-linux-gnu/luafar2l.so`.

**Reply by unxed** [reply]
Удаление помогло, больше не падает!

**Comment by unxed** [comment]
Итого, текущие исходники собираются так:
1. `sudo apt install lua5.1 liblua5.1-0`
2. Делаем `git clone` исходников far2l в home (если хотим клонировать в другое место, нужно будет поправить соответствующим образом первые две строчки в `Plugins/_luafar_/config.mak`), собираем
3. Кладем дистрибутив LuaFAR в `~/far2l/_build/install/Plugins/_luafar_`
4. В `_luafar_/src` делаем `make`
5. В `_luafar_/_lf4ed_/build` делаем `make embed`
6. В `_luafar_/_lfsearch_/build` делаем `make -fembed.mak`

**Reply by shmuz** [reply]
Большое спасибо! Мне повезло, что вы этим заинтересовались.

**Comment by unxed** [comment]
Теперь бы ещё домашнюю папку дерева исходников far2l задавать какой-нибудь переменной, чтоб без привязки к ~/far2l

**Comment by shmuz** [comment]
> Теперь бы ещё домашнюю папку дерева исходников far2l задавать какой-нибудь переменной

Попробуйте отредактировать одну или две верхние строки в `_luafar_/config.mak`.

**Reply by unxed** [reply]
А, точно :) Спасибо!

**Comment by spvkgn** [comment]
В far2l установленный в `/usr/lib/far2l` из ppa этот плагин работает? Собрал, положил в Plugins, но в меню F11 ничего не добавилось.
```
/usr/lib/far2l/Plugins/_luafar_
├── _lf4ed_
│  └── plug
│     ├── _plug_init.lua
│     ├── _usermenu.lua
│     ├── config.lua
│     ├── doc
│     │  ├── changelog.txt
│     │  ├── lf4ed_manual.chm
│     │  ├── lf4ed_manual.tsi
│     │  └── license.txt
│     ├── expression.lua
│     ├── lf4ed.lua
│     ├── lf4ed_e-x64.far-plug-wide
│     ├── lf4ed_eng.hlf
│     ├── lf4ed_eng.lng
│     ├── lf4ed_lang.lua
│     ├── lf4ed_lang.templ
│     ├── lf4ed_message.lua
│     ├── lf4ed_rus.hlf
│     ├── lf4ed_rus.lng
│     ├── scripts
│     │  ├── bracket.lua
│     │  ├── calc.lua
│     │  ├── del_trees.lua
│     │  ├── lf_rename.lua
│     │  ├── macrobrowser.lua
│     │  ├── multiline.lua
│     │  ├── test_sortlines.lua
│     │  └── utils_en.hlf
│     ├── sortdialog.lua
│     ├── sortlines.lua
│     └── wrap.lua
├── _lfsearch_
│  └── plug
│     ├── _plug_init.lua
│     ├── _usermenu.lua
│     ├── doc
│     │  ├── 3rd_party.txt
│     │  ├── changelog.txt
│     │  └── license.txt
│     ├── lfs_eng.hlf
│     ├── lfs_eng.lng
│     ├── lfs_engine.lua
│     ├── lfs_lang.lua
│     ├── lfs_lang.templ
│     ├── lfs_message.lua
│     ├── lfs_panels.lua
│     ├── lfs_rus.hlf
│     ├── lfs_rus.lng
│     ├── lfsearch.lua
│     ├── lfsearch_e-x64.far-plug-wide
│     ├── luarepl.lua
│     └── scripts
│        └── selftest.lua
└── luafar2l.so
```
```
$ ldd lf4ed_e-x64.far-plug-wide
	linux-vdso.so.1 (0x00007ffeabd8b000)
	../../luafar2l.so (0x00007f9084af3000)
	liblua5.1.so.0 => /usr/lib/x86_64-linux-gnu/liblua5.1.so.0 (0x00007f9084a99000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f90848a7000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f9084758000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f9084752000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9084b63000)

```

**Reply by shmuz** [reply]
Мне затруднительно дать ответ, поскольку предполагаю, что это связано с правами доступа/исполнения на Линуксе, которые я пока ещё не изучил. Приведенное вами дерево файлов выглядит вполне корректно, информация от ldd тоже.
Может быть Знатоки подскажут, почему работает, когда установлено под `$HOME`, но не работает под `/usr/lib/`

**Reply by shmuz** [reply]
@spvkgn , я сейчас (с помощью гугла) добился работы luafar-плагинов из-под `/usr/lib`
Для этого сделал следующее:
1. зашёл в папку `Plugins`
2. исполнил `sudo chmod -R 777 ./_luafar_`
3. перезапустил Фар

Это может быть в каком-то смысле неправильно, но уже шаг вперёд.

**Comment by unxed** [comment]
Вот сейчас собрал плагины, скопировал в /usr/lib/far2l/Plugins, всё заработало. Но у меня не из ppa, а самосборный .deb-пакет.

Кстати, из меню F11 плагин LuaFAR for Editor исчез (в редакторе, впрочем, всё ещё доступен).

**Reply by shmuz** [reply]
> Кстати, из меню F11 плагин LuaFAR for Editor исчез (в редакторе, впрочем, всё ещё доступен).

В этом плагине есть возможность добавления утилит пользователем. Эти добавляемые утилиты прописываются в файле `_usermenu.lua`. Те 2 утилиты, что показывались в меню в панелях, я закомментировал (смотрите в конце файла).
Поскольку в панелях не осталось ни одной утилиты, плагин не добавил себя в меню плагинов в панелях.

**Reply by shmuz** [reply]
@unxed , дополню:
1. Файл `_usermenu.lua` - это файл, поддерживаемый пользователем, туда можно что-то добавлять или наоборот.
2. Утилиты, которые добавляются в этом файле, не обязательно появляются в меню. Например,
`    AddToMenu ("e", nil, "Ctrl+Shift+F7", "scripts/editor_luacheck.lua")`
добавляет утилиту статической проверки Lua-файла из редактора по нажатию указанного кнопкосочетания.
В меню не появляется, поскольку 2-й параметр вызова - `nil`.

**Comment by shmuz** [comment]
Только что сделал коммит, в котором есть моя реализация макро-браузера.
Кто будет пробовать - обязательно сделайте предварительно бэкап файла `key_macros.ini`, т.к. это первая версия, и вполне возможны баги.

**Comment by shmuz** [comment]
1. Сейчас для того, чтобы собрать luafar2l.so и все плагины, достаточно вызвать `make` из папки `_luafar_`
2. Поскольку здесь обсуждение проекта LuaFAR-2L возможно является офтопиком, открыл таб **Discussions** на странице проекта. Приглашаются все.
